src/Runner.Worker/CompositeStepsRunner.cs

```diff
@@ -0,0 +1,249 @@
+                        var actionEnvironment = templateEvaluator.EvaluateStepEnvironment(actionStep.Action.Environment, step.ExecutionContext.ExpressionValues, step.ExecutionContext.ExpressionFunctions, VarUtil.EnvironmentVariableKeyComparer);
+                        foreach (var env in actionEnvironment)
+                        {
+                            envContext[env.Key] = new StringContextData(env.Value ?? string.Empty);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        // fail the step since there is an evaluate error.
+                        Trace.Info("Caught exception in Composite Steps Runner from expression for step.env");
+                        // evaluateStepEnvFailed = true;
+                        step.ExecutionContext.Error(ex);
+                        // CompleteStep(step, TaskResult.Failed);
+                    }
+                }
+
+                // We don't have to worry about the cancellation token stuff because that's handled by the composite action level (in the StepsRunner)
+
+                await RunStepAsync(step);
+
+                // TODO: Add compat for other types of steps.
+            }
+            // Completion Status handled by StepsRunner for the whole Composite Action Step
+        }
+
+        private async Task RunStepAsync(IStep step)
+        {
+            // Check to see if we can expand the display name
+            if (step is IActionRunner actionRunner &&
+                actionRunner.Stage == ActionRunStage.Main &&
+                actionRunner.TryEvaluateDisplayName(step.ExecutionContext.ExpressionValues, step.ExecutionContext))
+            {
+                step.ExecutionContext.UpdateTimelineRecordDisplayName(actionRunner.DisplayName);
+            }
+
+            // Start the step.
+            Trace.Info("Starting the step.");
+            step.ExecutionContext.Debug($"Starting: {step.DisplayName}");
+
+            // Set the timeout
+            var timeoutMinutes = 0;
+            var templateEvaluator = step.ExecutionContext.ToPipelineTemplateEvaluator();
+            try
+            {
+                timeoutMinutes = templateEvaluator.EvaluateStepTimeout(step.Timeout, step.ExecutionContext.ExpressionValues, step.ExecutionContext.ExpressionFunctions);
+            }
+            catch (Exception ex)
+            {
+                Trace.Info("An error occurred when attempting to determine the step timeout.");
+                Trace.Error(ex);
+                step.ExecutionContext.Error("An error occurred when attempting to determine the step timeout.");
+                step.ExecutionContext.Error(ex);
+            }
+            if (timeoutMinutes > 0)
+            {
+                var timeout = TimeSpan.FromMinutes(timeoutMinutes);
+                step.ExecutionContext.SetTimeout(timeout);
+            }
+
+#if OS_WINDOWS
+            try
```

Comment:
Do we need to do this? Wouldn't steps-runner have already attempted this?

Reply:
I guess technically a step could change it? Is that why we run it before each step? @TingluoHuang do you happen to remember?

Reply:
If we need to keep it here, it would be good to move into IOUtil.cs so we don't duplicate this lengthy code.

Reply:
+1 we don't really need this.

Reply:
(tl;dr from discussion, we should move this if step into IOUtil.cs)

Reply:
@ericsciple 

I don't actually can't abstract this away into IOUtil.cs because, 

`var p = HostContext.CreateService<IProcessInvoker>()` uses `RunnerService.HostContext` which I can't import into IOUtil.cs